File: observer/go/src/publisher.go
package src

import "time"

type Publisher struct {
	subscriptions []func()
}

func (p *Publisher) Start() {
	time.Sleep(time.Microsecond * 5)
	p.end()
}

func (p *Publisher) end() {
	p.notify()
}

func (p *Publisher) notify() {
	for _, s := range p.subscriptions {
		s()
	}
}

func (p *Publisher) Subscribe(f func()) int {
	p.subscriptions = append(p.subscriptions, f)

	return len(p.subscriptions) - 1
}

func (p *Publisher) Unsubscribe(i int) {
	p.subscriptions = append(p.subscriptions[:i], p.subscriptions[i+1:len(p.subscriptions)]...)
}

File: observer/scala/src/main/scala/Publisher.scala
package observer

class Publisher {
  var subscriptions: List[String => Unit] = List()

  def start(): Unit = {
    println("Publisher: Wasting time...")
    Thread.sleep(5000)
    end()
  }

  def end(): Unit = {
    println("Publisher: Notifying subscribers that I am done wasting time...")
    notify("Done")
  }

  def notify(d: String): Unit = {
    for (s <- subscriptions) s(d)
  }

  def subscribe(f: String => Unit): Int = {
    println("Publisher: Subscribing a subscriber...")
    subscriptions = subscriptions ++ List(f)
    subscriptions.length - 1
  }

  def unsubscribe(i: Int): Unit = {
    println("Publisher: Unsubscribing a subscriber...")
    subscriptions = subscriptions.patch(i, Nil, 1)
  }
}
